### DISCUSSION TITLE
Discussion about Specification pattern and it use in Checkout

### DISCUSSION THREAD

**Discussion by Ferror** [Sylius/Sylius#13807] [General]
## Introduction
Source PR: https://github.com/Sylius/Sylius/pull/13793

Extracted Specification code: https://gist.github.com/Ferror/4bf4fbd0cd6848bc66841725bfae92f9

## Questions/Ideas

1. Should a Specification pattern be created in the PR to be generic and reusable and treated as value-added or Order-Specific.
2. Should we use Specification implementation as Composite Tree (Dynamic Tree Approach in the gist) or Iterable (Iterable Approach in the gist)?

**Comment by vvasiloi** [comment]
1. I would go for a order-specific approach at first. Later, if necessary, it can be extracted and abstracted into a generic and reusable component.
2. I vote for the tree approach, but with a slightly different implementation. I would drop the `and` method and the abstract CompositeSpecification and use AndSpefication, OrSpecification and other specifications that will handle various compositions.

Example implementation: https://3v4l.org/sTLim
This allows indefinite nesting that will create a composite tree, for example the condition `X and Y or Z` will be represented as:
```php
$spec = new OrSpecification(
    new AndSpecification(
        new XSpecification(),
        new YSpecification(),
    ),
    new ZSpecification(),
);
```
